### ISSUE TITLE
RFE: automatically sign packages on build

### ISSUE THREAD

**Issue by pmatilai** [rpm-software-management/rpm#2678]
If we automatically signed all built packages, enabling enforcing signature checking by default (#1573) would be much less of an issue because it would no longer require --nosignature in common use-cases.

The gpg-agent stuff seems ill-suited for builds in non-interactive environments. I think rpm5 did something with keyutils(7), that may be something to look into.

**Comment by nwalfield** [comment]
A couple of thoughts:

  - In non-interactive environments, the secret key material should probably not be made available to the build infrastructure.  That means exactly something like gpg-agent (a daemon that provides a smartcard like interface) is needed.
  - How does rpm figure out what key to use for signing?

**Comment by pmatilai** [comment]
Oh, and part of the this "automation vision" here would be automatically generating that "local builds" key so that a person just wanting to build and install rpms for their own use basically doesn't need to learn the damnest thing about PGP as the first thing. And not finger-memorize --nosignature as something you normally use (against the background that rpm really, really needs to start requiring signatures by default)

**Comment by nwalfield** [comment]
> Oh, and part of the this "automation vision" here would be automatically generating that "local builds" key so that a person just wanting to build and install rpms for their own use basically doesn't need to learn the damnest thing about PGP as the first thing. And not finger-memorize --nosignature as something you normally use (against the background that rpm really, really needs to start requiring signatures by default)

Now I understand better, thanks.

For the case where the rpms are only going to be installed locally, it would be enough to have a local, user-specific rpm-specific key that is used for this type of thing.  If the key does not exist, then it is first created.  When the user wants to install the rpm, they first need to install the local, user-specific key, which could be done by `rpmkeys`.  `rpm` could even detect that the signature is from a local, user-specific key (perhaps by way of a notation added to the signature), and print a message telling the user how to install the key.  What do you think?

**Comment by pmatilai** [comment]
Yes, something along those lines. From the user POV, ideally one wouldn't even need the manual import (think something like auto-generated host-specific key on first boot), but that's likely too much of a sacrifice from other perspectives.

**Comment by nwalfield** [comment]
>From the user POV, ideally one wouldn't even need the manual import (think something like auto-generated host-specific key on first boot), but that's likely too much of a sacrifice from other perspectives.

But only root should have access to that key.  That's why I was thinking "user-specific".  In which case, I think rpm can't automatically import the key safely.

**Comment by pmatilai** [comment]
But, in this scenario the user is (or rather, has) root anyway, otherwise they wouldn't be able to install their packages in the first place. So it's this funny dance around how to best serve that case without sacrifing other scenarios. User-specific keys are the most obvious route of course, a host-specific key would solve something but be cumbersome/problematic in other ways. 
I'm just trying to think outside the box a bit, I don't have any concrete plan here.

**Comment by pmatilai** [comment]
And yes, because the user is likely root and expected to run rpm/dnf as such, expecting them to run one 'rpmkeys --import' command is not exactly unreasonable. It's largely about finding the right slot(s) to do stuff (such as print meaningful "do this" messages)

**Comment by stewartsmith** [comment]
I'll add some context on how we think about / do package signing in Amazon Linux (and for brevity, I'm going to just focus on AL2023 and beyond, and simplify things along the way) which may help inform decisions here.

Builders do not have access to the private signing keys. Nothing does. It's all in KMS. The keys were created in KMS, and nobody has **ever** seen the private key.

A (one to N instances of) a **separate** system to Koji works out what it should sign and with what key. This helps with things such as having an external audit log, and a simpler system to do an in-depth security review on. It also makes it **impossible** for any Koji deployment to accidentally go and [do something incredibly undesirable to the keys](https://docs.aws.amazon.com/kms/latest/developerguide/deleting-keys.html).

We haven't solved the scratch/local build signing problem, nor currently looked to. If I were to posit or review such a design for us doing it, I'd ensure the following constraints were true:
1. Anything not a final build in Koji is signed with a *different* key than what that final build would be signed for if it was queued for release. i.e. there is no way to have a random scratch/mock build to make it into any production system with a valid signature from the production keys.
2. Keys need to be **clearly** marked as for test and throwaway builds, and in such a way that it is possible for systems to alarm on any of these builds or keys making their way onto any production system.

In a previous work life, I worked on firmware that had the ability to do secure boot of the firmware stack. So that test keys were obvious, there was a standard set of test keys where the **private keys were in the public git repository** so that *everyone* knew what the test keys were.

**Comment by pmatilai** [comment]
> - rpmbuild always signs the packages it builds (unless explicitly disabled by config/cli switch)

On a related note, something @DemiMarie brought up in another ticket: perhaps a signature be actually a mandatory part of a v6 package. That seems like an attractive idea.

**Comment by pmatilai** [comment]
Righty, I guess this is starting to take some shape finally. Up to now I've been thinking of autosigning on build if %_openpgp_sign_id (think old %_gpg_name) is set, but this is problematic in many ways: signing takes place after the build has completed, and the signing asks for a passphrase at the end of four hours of build... and if you happen to be around, and mistype the password, you basically lost the build for no good reason. This is a pretty terrible user experience in every way 😆

So I think the right thing to do is to have rpm always setup a passwordless, rpm-specific key that all completed builds are signed with. If the user wants to resign it with something else like their own personal key later, that's no different to what they'd be doing now, and no matter what your setup was, builds aren't disrupted by any silly password questions that really don't belong in the build stage in the first place. 

With that, we get verifiable packages out of the gate for the local builds use-case. Distros and such signing their packages is an entirely different case, that we're not disrupting with this AFAICS.

**Comment by pmatilai** [comment]
So this part of the equation is rather simple then, AC:

- if %_openpgp_autosign_id macro is set, automatically sign packages at the end of a successful build using that key
- fail the build if signing fails

**Comment by marmarek** [comment]
The current approach looks okay, but I'll voice my concern in case it would be changed in a later iteration: avoid automatic key generation, and even more avoid automatically importing that key as trusted by rpm. This could easily result in the private key (already trusted by rpm there) leaking if somebody is not aware of all the details - for example by pushing a container to some registry where an rpm was built as part of the container build process.

**Comment by pmatilai** [comment]
@marmarek , feedback/discussion on the key generation/setup is most welcome in #3522 

We certainly cannot and will not auto-import any keys to rpm, that's just not a decision rpm can make on the users behalf.
